### ISSUE TITLE
Consul SD Config concurrent connection issues

### ISSUE THREAD

**Issue by roidelapluie** [prometheus/prometheus#7626]
hum, there is an additional problem: Consul team did add a limitation on the number of concurrent requests by default in 1.6.2.

So, actually, the problem is more complex:
 * on a Consul instance with very high concurrent connections allowed, waiting 10m is probably the best: requests are blocking and values updated as fast as possible, you probably have 2k opened connections all the time to the agent if you are watching all the services at the same time, updates are instant
 * On a consul instance with default limited connections (aka 200 in 1.6.3+), if you watch more than 200 services for instances, some connections will be silently cut by Consul agent in version < 1.8 (see https://github.com/hashicorp/consul/issues/7527), in version > 1.8, it would return HTTP 429 (that helps quite a lot to detect the issue instead of a raw connection closed)

In the latter case, even with 30s instead of 10m, you had an issue, you had no idea when you would get updates because you could be rate-limited and thus some updates would be never performed, but when lucky (assuming the number of watches is not too big compared to number of allowed connections), it kinda work.

Probably, to make it plug and play:
 * Use long blocking queries when enough connections are allowed
 * When not enough connections are allowed, perform quick calls (not blocking) every n seconds to get the data (but you loose instant notifications)

@roidelapluie the long polling works well in my case for instance, because I allowed more connections on my agents that the number of watches, but the change from 10 minutes to 30 seconds is probably problematic for other people.

A possibility would be to add a new parameter (set by default to 200, maybe even better that could be read from attribute `HTTPMaxConnsPerClient` in agent's address in `http://address_of_cousul_agent/v1/agent/self`).
* When we watch more than HTTPMaxConnsPerClient => switch to instant polling every n seconds
* If we watch less than that, stay on blocking queries

Any opinion on that?

_Originally posted by @pierresouchay in https://github.com/prometheus/consul_exporter/pull/176#issuecomment-661807898_

**Comment by roidelapluie** [comment]
I classify this a as a P1 bug for now because my understanding is that Prometheus v2.20.0 will silently only watch 200 services with default consul configuration.

**Comment by roidelapluie** [comment]
cc @beorn7 @simonpasquier 

I will have a look at this in the coming hours.

**Comment by brian-brazil** [comment]
> When we watch more than HTTPMaxConnsPerClient => switch to instant polling every n seconds

My understanding is that the whole point of how we currently work is to avoid polling, as that would cause even more problems on large consul setups with thousands of services. @iksaif @pierresouchay 

I'd suggest users set http_max_conns_per_client to value higher than the number of services they have. Designing a new way for Consul SD to work at scale will likely take a while, if it's even possible.

I don't think this is a P1, this will have been like this since January when 1.6.3 came out, it's a breakage on Consul's end, and there's a workaround. I don't think we should be potentially blocking releases due to 3rd party changes breaking our SDs.

**Comment by roidelapluie** [comment]
This is a P1 due to my investigation on the impact of the change in the current release. I am almost done.

**Comment by roidelapluie** [comment]
Here is the result of my investigation:

1. The errors are well visible in the logs and  sd_consul_rpc_failures_total is increasing when you set the limit.
2. While we could add extra parameters like: max_concurrent_connections, setting it would actually force us to disable polling for every service, otherwise the current polls would be done  again and again, with no space for other polls
3. Automatically disabling polling based on the available connection and the number of services is not desirable neither because if you have such a scale, you should have polling.

Proposed action for the 2.20 release: NONE

Proposed action for the master: when consul connection fail, look for error 429 and log a warning to the user about consul http_max_conns_per_client.

**Comment by roidelapluie** [comment]
Currently, we do not remove targets in the consul_sd when the watch is failing. This is something else to consider.

**Comment by brian-brazil** [comment]
> Currently, we do not remove targets in the consul_sd when the watch is failing. This is something else to consider.

In general if an SD is failing we should stick with the old targets. That way if there's a blip we don't stop scraping them.

**Comment by hbokh** [comment]
FWIW I arrived here because I upgraded Prometheus to v2.20.0 from v2.19.2 last Thursday.
In the Grafana dashboard "Prometheus 2.0 Overview" I saw the  amount of Consul RPC failures going up (`prometheus_sd_consul_rpc_failures_total`) since then.
"Downgraded" to v2.19.3 which seems to do fine so far.

**Comment by roidelapluie** [comment]
Do you see errors in the logs?

**Comment by hbokh** [comment]
> Do you see errors in the logs?

Yes, I saved some:of 'em

```console
level=error ts=2020-07-26T15:25:21.688Z caller=consul.go:487 component="discovery manager scrape" discovery=consul msg="Error refreshing service" service=mongo-primary tags= err="Get \"http://192.168.1.1:8500/v1/health/service/mongo-primary?index=170347&stale=&wait=600000ms\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)",
level=error ts=2020-07-26T15:25:21.677Z caller=consul.go:370 component="discovery manager scrape" discovery=consul msg="Error refreshing service list" err="Get \"http://192.168.1.1:8500/v1/catalog/services?index=170345&stale=&wait=600000ms\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)"
```

With v2.19.3 all clean since started.

Before going from v2.20.0 to v2.19.3 I also upped the limits for Consul:

```hcl
limits {
    http_max_conns_per_client = 500
}
```

**Comment by shamil** [comment]
I've exactly same issue, and getting same error. Downgrading to `2.19.2` solved the problem.

![image](https://user-images.githubusercontent.com/655447/88913663-16da6800-d26a-11ea-90bb-d89f4195901b.png)

**Comment by roidelapluie** [comment]
How many services do you have and what is your :

number of services
consul version
http_max_conns_per_client

thanks

**Comment by shamil** [comment]
@roidelapluie here is requested information

29 services 
consul version `1.7.5`
`http_max_conns_per_client` set to default (200)

I didn't have single issue before 2.20. I guess this some kind of regression that introduced in latest prometheus version

**Comment by roidelapluie** [comment]
This is a valid bug. See #7723 . However it should not affect the service discovery, it should work as expected, with a watch of still 30s effectively. It is just super annoying logs.

**Comment by shamil** [comment]
I don't think it's super annoying logs, those are real errors, since my other processes on the same instance which also connect to consul agent started to get connectivity issues after upgrading to `2.20`. Once reverted all back to normal...

**Comment by roidelapluie** [comment]
Yes these are real errors but the service discovery should still work as expected

**Comment by roidelapluie** [comment]
Maybe that is caused issues on consul end if the connections are not closed properly

**Comment by analytically** [comment]
Seeing this as well, 180 services, http_max_conns_per_client set to 10000, Consul 1.8.3, Prometheus 2.21.0.

**Comment by dgl** [comment]
We're also seeing a change in the behaviour going from Prometheus 2.16.0 to 2.21.0. There's actually multiple changes I see 
as relevant:
- The previously mentioned timeout changes (15086780017c2f7caf231e137b1ad9813eeb56c3, then 3a7120bc07f49d35f537bee6c81f5e6e849de0f8)
- 9b9c58aea8e0043bb944961fb776b48ac602c7f9 which changed from using the service endpoint to the consul health API endpoint.

In particular I suspect the switch to the health endpoint interacts poorly with network events, i.e. if there are any network issues health of many services may change, resulting in potentially a lot of updates, with following reconnections, when the network is potentially having problems.

One hope is in Consul 1.9 there is "Streaming" (https://www.hashicorp.com/blog/announcing-hashicorp-consul-1-9), which could help reduce the need for so many reconnections.

**Comment by analytically** [comment]
I resolved this issue by moving from AWS ALB (in front of Consul) to NLB.

**Comment by jakubgs** [comment]
My question would be, why is the query using something like `wait=600000ms` - or in my case `wait=120000ms`? Why is it waiting for the query instead of making it right away?

**Comment by jasonk** [comment]
It's waiting for the query because it's a long poll.  It already ran it and got a response, the query that it sends with `wait=128000ms` also includes an `index=` parameter.  This query is basically saying it already has the data up to that index and now wants to be notified immediately of any changes to that service so it's going to send a query that waits for up to 2 minutes for a change, then if that times out it will make the same query again and again and again until something does change.

**Comment by patsevanton** [comment]
Will setting the following parameter in prometheus help?
```
consul_sd_config:
  refresh_interval: 60 # default = 30s
```

**Comment by patsevanton** [comment]
We are fixed by:
```
    ingress:
      enabled: true
      hostname: consul-url
      annotations:
        kubernetes.io/ingress.class: nginx
        nginx.ingress.kubernetes.io/proxy-connect-timeout: "300"
        nginx.ingress.kubernetes.io/proxy-read-timeout: "300"
        nginx.ingress.kubernetes.io/proxy-send-timeout: "300"
```

**Comment by krajorama** [comment]
Hello from the bug scrub: we are still looking for somebody to propose a fix or alternatively a documentation PR as the issue is relevant. Keeping open.

**Comment by WojciechKuk** [comment]
I still observe this problem having > 2500 services in consul - prometheus creates 2500 connections without any rate-limiting.
of course refresh_interval will not help here - it's silence interval between DDoSs, not the interval between individual connections.
